### PR TITLE
[CDAP-17347] Don't show stack trace in log message when transform fails

### PR DESCRIPTION
**Pull request by rmstar**

### CODE REVIEW

**Comment by chtyim** [comment]
Then how the user see the stack trace to debug the python error?

**Comment by rmstar** [comment]
@chtyim an error message would still be displayed, just not a full stack trace.

Example:
```
def transform(record, emitter, context):
    if (record.id < 2):  # error: should be record['id'] instead of record.id
      emitter.emit({
        'id': 10,
        'firstname': 'f',
        'lastname': 'l'
      })
    else:
      emitter.emit(record)
```
In this case error msg looks like this:
Screen Shot 2021-02-22 at 10 04 21 PM

**Comment by rmstar** [comment]
Without the fix, you'd see this: 
Screen Shot 2021-02-22 at 12 55 26 PM

**Comment by rmstar** [comment]
Another thing to note is that with the fix you do see a stack trace in the raw logs like this:
```
2021-02-22 13:04:15,392 - ERROR [Executor task launch worker for task 0:o.a.s.u.Utils@91] - Aborting task
java.lang.IllegalArgumentException: Encountered an error in the Python code.
AttributeError("'java.util.HashMap' object has no attribute 'id'",)
	at io.cdap.plugin.python.transform.JythonPythonExecutor.transform(JythonPythonExecutor.java:84) ~[1614027848203-0/:na]
	at io.cdap.plugin.python.transform.PythonEvaluator.transform(PythonEvaluator.java:265) ~[1614027848203-0/:na]
	at io.cdap.plugin.python.transform.PythonEvaluator.transform(PythonEvaluator.java:61) ~[1614027848203-0/:na]
	at io.cdap.cdap.etl.common.plugin.WrappedTransform.lambda$transform$5(WrappedTransform.java:90) ~[cdap-etl-core-6.4.0-SNAPSHOT.jar:na]
	at io.cdap.cdap.etl.common.plugin.Caller$1.call(Caller.java:30) ~[cdap-etl-core-6.4.0-SNAPSHOT.jar:na]
	at io.cdap.cdap.etl.common.plugin.WrappedTransform.transform(WrappedTransform.java:89) ~[cdap-etl-core-6.4.0-SNAPSHOT.jar:na]
	at io.cdap.cdap.etl.common.TrackedTransform.transform(TrackedTransform.java:74) ~[cdap-etl-core-6.4.0-SNAPSHOT.jar:na]
	at io.cdap.cdap.etl.spark.function.TransformFunction.call(TransformFunction.java:50) ~[hydrator-spark-core2_2.11-6.4.0-SNAPSHOT.jar:na]
	at io.cdap.cdap.etl.spark.Compat$FlatMapAdapter.call(Compat.java:126) ~[hydrator-spark-core2_2.11-6.4.0-SNAPSHOT.jar:na]
	at org.apache.spark.api.java.JavaRDDLike$$anonfun$fn$1$1.apply(JavaRDDLike.scala:125) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.api.java.JavaRDDLike$$anonfun$fn$1$1.apply(JavaRDDLike.scala:125) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:434) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:439) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:439) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1$$anonfun$12$$anonfun$apply$4.apply$mcV$sp(PairRDDFunctions.scala:1123) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1$$anonfun$12$$anonfun$apply$4.apply(PairRDDFunctions.scala:1123) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1$$anonfun$12$$anonfun$apply$4.apply(PairRDDFunctions.scala:1123) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1356) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1$$anonfun$12.apply(PairRDDFunctions.scala:1131) [io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1$$anonfun$12.apply(PairRDDFunctions.scala:1102) [io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87) [io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.scheduler.Task.run(Task.scala:100) [io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:325) [io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_221]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_221]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_221]
```

File: src/main/java/io/cdap/plugin/python/transform/JythonPythonExecutor.java [hunk 1]
```diff
@@ -83,19 +81,10 @@ public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter,
       Py.runCode(compiledScript, interpreter.getLocals(), interpreter.getLocals());
 
     } catch (PyException e) {
-      // Put stack trace as the exception message, because otherwise the information from PyException is lost.
-      // PyException only exposes the actual cause (Python stack trace) if printStackTrace() is called on it.
-      throw new IllegalArgumentException("Could not transform input.\n" + getStackTrace(e));
+      throw new IllegalArgumentException("Encountered an error in the Python code.\n" + e.value);
```
Comment (chtyim) [inline src/main/java/io/cdap/plugin/python/transform/JythonPythonExecutor.java]: Isn't the `e` should be passed as the root cause?

**Comment by chtyim** [comment]
Also, without the python stacktrace, it would be difficult to debug bugs in python code.

File: src/main/java/io/cdap/plugin/python/transform/JythonPythonExecutor.java (hunk 1 above)
Comment (chtyim) [reply src/main/java/io/cdap/plugin/python/transform/JythonPythonExecutor.java]: Should be just `throw new IllegalArgumentException("... some message... ", e);`

**Comment by rmstar** [comment]
> Also, without the python stacktrace, it would be difficult to debug bugs in python code.

Would it be better to print the python stack trace at WARN level in raw logs? So the stacktrace would show up in advanced/raw logs.

**Comment by chtyim** [comment]
If you keep the original exception as the root cause, it will be shown in the raw log, isn't it? Why need a separate log line?

File: src/main/java/io/cdap/plugin/python/transform/JythonPythonExecutor.java (hunk 1 above)
Comment (rmstar) [reply src/main/java/io/cdap/plugin/python/transform/JythonPythonExecutor.java]: PyException is weird. `e.getCause()` returns null.
Screen Shot 2021-02-22 at 10 36 31 PM

File: src/main/java/io/cdap/plugin/python/transform/JythonPythonExecutor.java [hunk 2]
```diff
@@ -83,19 +85,11 @@ public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter,
       Py.runCode(compiledScript, interpreter.getLocals(), interpreter.getLocals());
 
     } catch (PyException e) {
-      // Put stack trace as the exception message, because otherwise the information from PyException is lost.
-      // PyException only exposes the actual cause (Python stack trace) if printStackTrace() is called on it.
-      throw new IllegalArgumentException("Could not transform input.\n" + getStackTrace(e));
+      LOGGER.warn("Transform failed", new Exception(e.toString()));
```
Comment (chtyim) [inline src/main/java/io/cdap/plugin/python/transform/JythonPythonExecutor.java]: Why log here?

Comment (rmstar) [reply src/main/java/io/cdap/plugin/python/transform/JythonPythonExecutor.java]: This msg log ensures that the python stacktrace shows up in raw logs
```
2021-02-23 11:50:07,322 - WARN  [Executor task launch worker for task 0:i.c.p.p.t.JythonPythonExecutor@88] - Transform failed
java.lang.Exception: Traceback (most recent call last):
  File "<script>", line 25, in <module>
  File "<script>", line 16, in transform
AttributeError: 'java.util.HashMap' object has no attribute 'id'

	at io.cdap.plugin.python.transform.JythonPythonExecutor.transform(JythonPythonExecutor.java:88) [1614109801877-0/:na]
	at io.cdap.plugin.python.transform.PythonEvaluator.transform(PythonEvaluator.java:265) [1614109801877-0/:na]
	at io.cdap.plugin.python.transform.PythonEvaluator.transform(PythonEvaluator.java:61) [1614109801877-0/:na]
	at io.cdap.cdap.etl.common.plugin.WrappedTransform.lambda$transform$5(WrappedTransform.java:90) [cdap-etl-core-6.4.0-SNAPSHOT.jar:na]
	at io.cdap.cdap.etl.common.plugin.Caller$1.call(Caller.java:30) ~[cdap-etl-core-6.4.0-SNAPSHOT.jar:na]
	at io.cdap.cdap.etl.common.plugin.WrappedTransform.transform(WrappedTransform.java:89) [cdap-etl-core-6.4.0-SNAPSHOT.jar:na]
	at io.cdap.cdap.etl.common.TrackedTransform.transform(TrackedTransform.java:74) ~[cdap-etl-core-6.4.0-SNAPSHOT.jar:na]
	at io.cdap.cdap.etl.spark.function.TransformFunction.call(TransformFunction.java:50) ~[hydrator-spark-core2_2.11-6.4.0-SNAPSHOT.jar:na]
	at io.cdap.cdap.etl.spark.Compat$FlatMapAdapter.call(Compat.java:126) ~[hydrator-spark-core2_2.11-6.4.0-SNAPSHOT.jar:na]
	at org.apache.spark.api.java.JavaRDDLike$$anonfun$fn$1$1.apply(JavaRDDLike.scala:125) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.api.java.JavaRDDLike$$anonfun$fn$1$1.apply(JavaRDDLike.scala:125) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:434) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:439) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:439) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1$$anonfun$12$$anonfun$apply$4.apply$mcV$sp(PairRDDFunctions.scala:1123) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1$$anonfun$12$$anonfun$apply$4.apply(PairRDDFunctions.scala:1123) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1$$anonfun$12$$anonfun$apply$4.apply(PairRDDFunctions.scala:1123) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1356) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1$$anonfun$12.apply(PairRDDFunctions.scala:1131) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1$$anonfun$12.apply(PairRDDFunctions.scala:1102) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.scheduler.Task.run(Task.scala:100) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:325) ~[io.cdap.cdap.spark-assembly-2.1.3.jar:na]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_221]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_221]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_221]
```

And log viewer looks like this: 
Screen Shot 2021-02-23 at 11 51 50 AM

**Review by chtyim — APPROVED** [review]

**Pull request merged**